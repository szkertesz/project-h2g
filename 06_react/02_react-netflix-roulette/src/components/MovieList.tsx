import { useAppDispatch, useAppSelector } from '../app/hooks';
import { fetchMovies, selectAllMovies } from '../features/movies/moviesSlice';
import { useEffect } from 'react';
import MovieItem from './MovieItem';
import { Movie } from '../interfaces/Movie';
import classes from './MovieList.module.scss';


// import { useAppDispatch, useAppSelector } from '../app/hooks';
// RTKQ
// import { useGetDataQuery } from '../features/api/apiSlice';

function MovieList() {
    // bare react code
    // const [isLoading, setIsLoading] = useState(false);
    // const [movieData, setMovieData] = useState<Movie[]>([]);

    // const getData = async () => {
    //     try {
    //         const response = await fetch('movies.json', {
    //             headers: {
    //                 'Content-Type': 'application/json',
    //                 Accept: 'application/json',
    //             },
    //         });
    //         if (!response.ok) {
    //             throw new Error(response.statusText);
    //         }
    //         const data = await response.json();
    //         const moviesData = data.movies;
    //         setMovieData(moviesData);
    //         setIsLoading(false);
    //     } catch (error) {
    //         if (error instanceof Error) {
    //             console.warn(String(error.message));
    //         }
    //     }
    // };

    // useEffect(() => {
    //     getData();
    // }, []);

    // if(isLoading) {
    //     return(
    //         <section>
    //             <h2 className='visually-hidden'>Loading State</h2>
    //             <p>Loading...</p>
    //         </section>
    //     )
    // }

    // RTKQ code
    // const { data, isFetching } = useGetDataQuery();
    // const movieData = data;

    // if (isFetching) return <section>Loading...</section>;
    // if (!data) return <section><p>Can't get any data :-/</p></section>;
    const dispatch = useAppDispatch();
    const movies = useAppSelector(selectAllMovies);
    const moviesStatus = useAppSelector(state => state.movies.status);
    const moviesError = useAppSelector(state => state.movies.error);

    useEffect(() => {
        if (moviesStatus === 'idle') {
            dispatch(fetchMovies())
        }
    }, [moviesStatus, dispatch]);

    if (moviesStatus === 'loading') {
        return (
            <section>
                <p>Loading...</p>
        </section>
            )
    } else if (moviesStatus === 'succeeded') {
        return (
            <section>
                <h2 className='visually-hidden'>Results</h2>
                <p className={classes.info}>
                    <span>{movies.length}</span> movies found
                </p>
                <ul className={classes.results}>
                    {movies.map((movie: Movie) => {
                        return (
                            <li key={movie.id}>
                                <MovieItem movieInfo={movie} />
                            </li>
                        );
                    })}
                </ul>
            </section>
        );
    } else {
        return (
            <section>
                <p>{ moviesError }</p>
            </section>
        )
    }
}

export default MovieList;
