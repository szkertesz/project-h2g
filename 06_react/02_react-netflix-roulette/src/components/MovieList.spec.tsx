import React from 'react';
import { Provider } from 'react-redux';
import renderer, {
    ReactTestInstance,
    ReactTestRenderer,
} from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import type { RootState } from '../app/store';

import MovieList from './MovieList';
import { Store, AnyAction } from '@reduxjs/toolkit';
import { testData } from '../testing/test-data-movies';
import { MemoryRouter } from 'react-router-dom';

const mockStore = configureStore<RootState>([]);

describe('MovieList', () => {
    let store: Store<any, AnyAction>;
    let component: ReactTestRenderer;
    let root: ReactTestInstance;

    beforeEach(() => {
        store = mockStore({
            movies: {
                data: testData.movies.data,
                status: 'succeeded',
                error: null,
                searchedMovie: '',
                sortOptions: {
                    sortCriterion: '',
                    sortOrder: '',
                },
                filterOptions: {
                    genre: '',
                },
            },
        });

        component = renderer.create(
            <Provider store={store}>
                <MemoryRouter>
                    <MovieList />
                </MemoryRouter>
            </Provider>
        );
    });
    // the test suite
    it('should render component with given movies data', () => {
        expect(component.toJSON()).toMatchSnapshot(); // the assertion
    });
});
