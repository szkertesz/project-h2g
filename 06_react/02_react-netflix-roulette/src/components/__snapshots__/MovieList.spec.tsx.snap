// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList should render component with given movies data 1`] = `
<section>
  <h2
    className="visually-hidden"
  >
    Results
  </h2>
  <p
    className="info"
  >
    <span>
      3
    </span>
     movies found
  </p>
  <ul
    className="results"
  >
    <li>
      <article
        className="movie-item"
      >
        <div
          className="movie-item__text"
        >
          <div>
            <h3
              className="movie-item__title"
            >
              Fifty Shades Freed
            </h3>
            <ul
              className="genre-list"
            >
              <li>
                Drama, 
              </li>
              <li>
                Romance
              </li>
            </ul>
          </div>
          <time
            className="movie-item__release"
            dateTime="2018"
          >
            2018
          </time>
        </div>
        <a
          className="movie-item__poster"
          href="/movie/337167"
          onClick={[Function]}
        >
          <img
            alt="poster"
            src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
          />
        </a>
        <div
          className="movie-item__actions"
        >
          <button
            className="button 
            button--default
            "
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="button 
            button--default
            "
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </article>
    </li>
    <li>
      <article
        className="movie-item"
      >
        <div
          className="movie-item__text"
        >
          <div>
            <h3
              className="movie-item__title"
            >
              Zootopia
            </h3>
            <ul
              className="genre-list"
            >
              <li>
                Animation, 
              </li>
              <li>
                Adventure, 
              </li>
              <li>
                Family, 
              </li>
              <li>
                Comedy
              </li>
            </ul>
          </div>
          <time
            className="movie-item__release"
            dateTime="2016"
          >
            2016
          </time>
        </div>
        <a
          className="movie-item__poster"
          href="/movie/269149"
          onClick={[Function]}
        >
          <img
            alt="poster"
            src="https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg"
          />
        </a>
        <div
          className="movie-item__actions"
        >
          <button
            className="button 
            button--default
            "
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="button 
            button--default
            "
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </article>
    </li>
    <li>
      <article
        className="movie-item"
      >
        <div
          className="movie-item__text"
        >
          <div>
            <h3
              className="movie-item__title"
            >
              Star Wars: The Last Jedi
            </h3>
            <ul
              className="genre-list"
            >
              <li>
                Fantasy, 
              </li>
              <li>
                Adventure, 
              </li>
              <li>
                Science Fiction
              </li>
            </ul>
          </div>
          <time
            className="movie-item__release"
            dateTime="2017"
          >
            2017
          </time>
        </div>
        <a
          className="movie-item__poster"
          href="/movie/181808"
          onClick={[Function]}
        >
          <img
            alt="poster"
            src="https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg"
          />
        </a>
        <div
          className="movie-item__actions"
        >
          <button
            className="button 
            button--default
            "
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="button 
            button--default
            "
            onClick={[Function]}
          >
            Edit
          </button>
        </div>
      </article>
    </li>
  </ul>
</section>
`;
